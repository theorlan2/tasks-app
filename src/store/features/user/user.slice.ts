import {
  createSelector,
  createEntityAdapter,
  EntityState,
  EntityId,
} from "@reduxjs/toolkit";

import { apiSlice } from "@/store/features/api/api.slice";

import { UserI } from "@/types/user/user.model";

type LocalUser = Omit<UserI, "id"> & { id: EntityId };

const usersAdapter = createEntityAdapter<LocalUser>();
const initialState = usersAdapter.getInitialState();

export const apiSliceWithUsers = apiSlice.injectEndpoints({
  endpoints: (builder) => ({
    getUsers: builder.query<EntityState<LocalUser, string>, void>({
      query: () => "/elements",
      transformResponse(res: LocalUser[]) {
        // Create a normalized state object containing all the user items
        return res;
      },
    }),
  }),
});

export const { useGetUsersQuery } = apiSliceWithUsers;

// Calling `someEndpoint.select(someArg)` generates a new selector that will return
// the query result object for a query with those parameters.
// To generate a selector for a specific query argument, call `select(theQueryArg)`.
// In this case, the users query has no params, so we don't pass anything to select()
export const selectUsersResult = apiSliceWithUsers.endpoints.getUsers.select();
const selectUsersData = createSelector(
  selectUsersResult,
  // Fall back to the empty entity state if no response yet.
  (result) => result.data ?? initialState,
);

export const { selectAll: selectAllUsers, selectById: selectUserById } =
  usersAdapter.getSelectors(selectUsersData);
