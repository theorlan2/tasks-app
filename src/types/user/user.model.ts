import { GenericEntity } from "@/types/shared/shared.model";

export interface UserI<T = number | string> extends GenericEntity<T> {
  name: string;
  avatar: string;
}
