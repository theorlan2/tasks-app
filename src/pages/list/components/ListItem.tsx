import { UserI } from "@/types/user/user.model";

type Props = {
  data: UserI;
};

const ListItem = ({ data }: Props) => {
  return (
    <div className=" flex flex-col-reverse gap-2 sm:flex-row relative justify-between rounded-md p-3 text-sm/6 transition border border-gray-200 hover:bg-gray-400/20 dark:hover:bg-white/10 my-2">
      <div className="flex flex-col w-full ">
        <p className="font-semibold text-lg text-black dark:text-white mb-0">
          {data.name}
        </p>
        <div className="w-full border-b  border-gray-100 dark:border-gray-50/30 my-2"></div>
        <span className="text-xs text-gray-500 dark:text-white ">
          {data.createdAt
            ? new Date(data.createdAt).toLocaleString()
            : "No date"}
        </span>
      </div>
      <div className="w-full sm:w-9 flex items-center ">
        <i className="ri-account-circle-line text-4xl "></i>
        {/* <img className="max-w-9" src={data.avatar} alt="Avatar" /> */}
      </div>
    </div>
  );
};
export default ListItem;
