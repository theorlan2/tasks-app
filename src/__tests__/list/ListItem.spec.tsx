import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";

import { UserI } from "@/types/user/user.model";

import ListItem from "@/pages/list/components/ListItem";

describe("ListItem Component", () => {
  const mockUser: UserI = {
    id: "1",
    name: "John Doe",
    createdAt: new Date().toUTCString(),
    avatar: "",
  };

  it("should render correctly", () => {
    const component = render(<ListItem data={mockUser} />);
    expect(component.container).toBeDefined();
  });

  it("should display the user name", () => {
    const { getByText } = render(<ListItem data={mockUser} />);

    expect(getByText(mockUser.name)).toBeTruthy();
  });

  it("should display the creation date", () => {
    const { getByText } = render(<ListItem data={mockUser} />);

    expect(
      getByText(new Date(mockUser.createdAt || "").toLocaleString()),
    ).toBeTruthy();
  });
});
