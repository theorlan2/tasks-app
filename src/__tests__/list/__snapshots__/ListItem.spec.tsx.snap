// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListItem Component > should render correctly 1`] = `
<div>
  <div
    class=" flex flex-col-reverse gap-2 sm:flex-row relative justify-between rounded-md p-3 text-sm/6 transition border border-gray-200 hover:bg-gray-400/20 dark:hover:bg-white/10 my-2"
  >
    <div
      class="flex flex-col w-full "
    >
      <p
        class="font-semibold text-lg text-black dark:text-white mb-0"
      >
        John Doe
      </p>
      <div
        class="w-full border-b  border-gray-100 dark:border-gray-50/30 my-2"
      />
      <span
        class="text-xs text-gray-500 dark:text-white "
      >
        1/23/2025, 8:34:08 PM
      </span>
    </div>
    <div
      class="w-full sm:w-9 flex items-center "
    >
      <i
        class="ri-account-circle-line text-4xl "
      />
    </div>
  </div>
</div>
`;
